src/shell/signals.c

```diff
@@ -25,15 +25,66 @@
 #include <signal.h>
 #include <flux/core.h>
 #include <flux/shell.h>
+#include <assert.h>
 
 #include "internal.h"
 #include "builtins.h"
 
+static inline void signal_exit (int signum)
+{
+    /* 128 + signum is standard bash exit code */
+    exit (128 + signum);
+}
+
 static void signal_cb (flux_reactor_t *r, flux_watcher_t *w,
                        int revents, void *arg)
 {
     flux_shell_t *shell = arg;
     int sig = flux_signal_watcher_get_signum (w);
+    int taskcount, ntasks;
+
+    /* Possibly impossible, but just in case */
+    if (!shell->tasks)
+        signal_exit (sig);
+
+    taskcount = zlist_size (shell->tasks);
+
+    if (flux_shell_info_unpack (shell,
+                                "{s:i}",
+                                "ntasks", &ntasks) < 0) {
+        shell_log_errno ("failed to unpack shell info");
+        signal_exit (sig);
+    }
+
+    /* Very rare and unlikely race, signal received before all tasks
+     * started.  Hypothetically could be seen with something like:
+     *
+     * jobid=`flux job submit --wait-event=start foo.sh`
+     * flux job raise --type=foo --severity=0 $jobid
+     *
+     * i.e. SIGTERM sent to job/shell immediately after starting.
+     * This should not be seen under normal operating circumstances.
+     *
+     * This race can lead to unexpected errors in task launching due
+     * to the signal interrupting initial setup.  Since this situation
+     * is extremely hard to hit and behavior is unexpected, we elect
+     * to log the situation and exit.
+     *
+     * Note that we could have chosen to call flux_shell_killall() on
+     * all the currently running tasks with the desired signal.
+     * However, given how racy this situation is, we can't be sure the
+     * tasks have setup their own signal handlers yet.  So we elect to
+     * just kill those tasks and exit.  The presumption is that if
+     * this situation were ever hit in normal operating conditions, it
+     * is because the job was messed up and killing the job is the
+     * desired result.
+     */
+    if (taskcount < ntasks) {
+        shell_log_errno ("signal %d caught before task(s) setup, exiting", sig);
```

Comment:
Hm, instead of all this, I wonder if it would be cleaner to just call `shell_die (128+sig, "signal %d sent before tasks started")`, since that has the intended effect without having to unpack shell info while handling a signal, etc.

Reply:
Yeah, as I thought about the circumstances more, there were a lot of "ohh but we should really check for ..."

Certainly could do `shell_die()` instead of my `signal_exit()` call though, didn't think to do that.